### PR TITLE
SAA-1064 new endpoint to support adding prisoners to waiting lists for activities

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+@Transactional(readOnly = true)
+class WaitingListService(
+  private val scheduleRepository: ActivityScheduleRepository,
+  private val waitingListRepository: WaitingListRepository,
+  private val prisonApiClient: PrisonApiClient,
+) {
+
+  @Transactional
+  @PreAuthorize("hasAnyRole('ACTIVITY_HUB', 'ACTIVITY_HUB_LEAD', 'ACTIVITY_ADMIN')")
+  fun addPrisoner(prisonCode: String, request: WaitingListCreateRequest, createdBy: String) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:30]: FYI - this method is incomplete.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt [hunk 2]
```diff
@@ -40,6 +41,11 @@ class WaitingList(
 
   val createdBy: String,
 ) {
+
+  @ManyToOne
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:45]: Not sure if we really need activity as well.  Here now.  Can remove if we feel this is overkill.

File: src/main/resources/migrations/common/V2023.08.02__alter_waiting_list.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE waiting_list ADD FOREIGN KEY(activity_id) REFERENCES activity(activity_id);
```
Comment (matt-crowson-moj) [inline src/main/resources/migrations/common/V2023.08.02__alter_waiting_list.sql:1]: Missed of the original table, grrr.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt [hunk 4]
```diff
@@ -91,6 +91,10 @@ data class ActivitySchedule(
   @Fetch(FetchMode.SUBSELECT)
   private val slots: MutableList<ActivityScheduleSlot> = mutableListOf()
 
+  @OneToMany(mappedBy = "activitySchedule", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
+  @Fetch(FetchMode.SUBSELECT)
+  private val waiting: MutableList<WaitingList> = mutableListOf()
+
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:97]: Worth a discussion anyway .. as this is populated by many API endpoints - called a lot.  And as its a chance to do this differently than instances, or allocations, might be a good opportunity?

[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:97]: I did actually originally have this all being done in the service direct.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/WaitingListApplicationRequest.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    example = "PENDING",
+  )
+  @field:NotNull(message = "Status must be supplied")
+  val status: Status?,
+)
+
+enum class Status(val waitingListStatus: WaitingListStatus) {
+  APPROVED(WaitingListStatus.APPROVED),
+  DECLINED(WaitingListStatus.DECLINED),
+  PENDING(WaitingListStatus.PENDING),
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/WaitingListApplicationRequest.kt:67]: No.  This is what the front end sends/supports.  I admit it does look odd.  We can just use them all.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/ActivityScheduleRepository.kt [hunk 6]
```diff
@@ -13,4 +15,11 @@ interface ActivityScheduleRepository : JpaRepository<ActivitySchedule, Long>, Ac
 
   @EntityGraph(attributePaths = ["instances"], type = EntityGraph.EntityGraphType.LOAD)
   fun getAllByActivity(activity: Activity): List<ActivitySchedule>
+
+  @Query(
+    value = """
+    from ActivitySchedule s where s.activityScheduleId = :id and s.activity.prisonCode = :prisonCode
+  """,
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/ActivityScheduleRepository.kt:22]: It is but you still want to make sure it is for the same prison.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/InmateDetailFixture.kt [hunk 7]
```diff
@@ -14,7 +14,7 @@ object InmateDetailFixture {
     activeFlag: Boolean = true,
     inOutStatus: String = "IN",
     status: String = "IN",
-    bookingId: Long = 900001,
+    bookingId: Long? = 900001,
     bookingNo: String = "BK01",
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/InmateDetailFixture.kt:18]: Can `bookingId` be null?

**Review by timharrison-moj — APPROVED** [review]
LGTM. Small question only.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/InmateDetailFixture.kt (hunk 7 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/InmateDetailFixture.kt:18]: This is down to the prison api return type.  It allows a nullable booking id, grrr.  Doing this allows me to test and handle it (even if it is an edge case).
